src/clients/c++/experimental_api_v2/perf_client/model_parser.cc

```diff
@@ -0,0 +1,297 @@
+    auto it = inputs_->emplace(input.name(), ModelTensor()).first;
+    it->second.name_ = input.name();
+    it->second.datatype_ = input.datatype();
+    bool is_dynamic = false;
+    for (const auto dim : input.shape()) {
+      if (dim == -1) {
+        is_dynamic = true;
+      }
+      it->second.shape_.push_back(dim);
+    }
+
+    if (is_dynamic) {
+      const auto user_shape_it = input_shapes.find(input.name());
+      if (user_shape_it != input_shapes.end()) {
+        // Update the default shape to be used.
+        it->second.shape_.clear();
+        for (const auto dim : user_shape_it->second) {
+          it->second.shape_.push_back(dim);
+        }
+      }
+    }
+  }
+
+  // Check whether the tensor is shape tensor or not from config.
+  for (const auto& input_config : config.input()) {
+    const auto& itr = inputs_->find(input_config.name());
+    if (itr == inputs_->end()) {
+      return nic::Error(
+          "no metadata found for input tensor " + input_config.name());
+    }
+    itr->second.is_shape_tensor_ = input_config.is_shape_tensor();
+  }
+
+  // Get the information about outputs from metadata
+  for (const auto& output : metadata.outputs()) {
+    auto it = outputs_->emplace(output.name(), ModelTensor()).first;
+    it->second.name_ = output.name();
+    it->second.datatype_ = output.datatype();
+    for (const auto dim : output.shape()) {
+      it->second.shape_.push_back(dim);
+    }
+  }
+
+  // Check whether the tensor is shape tensor or not from config.
+  for (const auto& output_config : config.output()) {
+    const auto& itr = outputs_->find(output_config.name());
+    if (itr == outputs_->end()) {
+      return nic::Error(
+          "no metadata found for output tensor " + output_config.name());
+    }
+    itr->second.is_shape_tensor_ = output_config.is_shape_tensor();
+  }
+  return nic::Error::Success;
+}
+
+
+nic::Error
+ModelParser::GetEnsembleSchedulerType(
+    const ni::ModelConfig& config, const std::string& model_version,
+    std::unique_ptr<TritonClientWrapper>& client_wrapper, bool* is_sequential)
+{
+  if (config.has_sequence_batching()) {
+    *is_sequential = true;
+  }
+
+  if (config.platform() == "ensemble") {
+    for (const auto& step : config.ensemble_scheduling().step()) {
+      ni::ModelConfigResponse model_config;
+      std::string step_version;
+      if (step.model_version() != -1) {
+        step_version = std::to_string(step.model_version());
+      }
+      (*composing_models_map_)[config.name()].emplace(
+          step.model_name(), step_version);
+      RETURN_IF_ERROR(client_wrapper->ModelConfig(
+          &model_config, step.model_name(), step_version));
+      RETURN_IF_ERROR(GetEnsembleSchedulerType(
+          model_config.config(), step_version, client_wrapper, is_sequential));
+    }
+  }
+
+  return nic::Error::Success;
+}
+
+nic::Error
+ModelParser::Init(
+    const rapidjson::Document& metadata, const rapidjson::Document& config,
```

Comment:
I think `config` can be coverted to corresponding protobuf object, and vice versa. Can just convert it and call the other Init() function?

Reply:
We use config just to detect scheduler type and detecting which tensors are shape tensor. I believe it would be wasteful to convert config DOM into protobuf then extracting those values instead of directly extracting those values from DOM. If we were to call the other Init() function we would have to convert metadata to protobuf as well.